### PR TITLE
Overhauled the time syncing algorithm

### PR DESCRIPTION
**Pull request by V-Fries**
Client can now differentiate latency and bad system time sync

Client sends a `request_time_sync` event to the server:
```
Argument:
time1: Date.now()
```
Server receives the event and sends a `time_sync` event to the client:
```
Argument:
{
    time1: time1 (Copy of the argument given by the client),
    time2: time.time()
}
```
Client receives the event and can now calculate how long the server took to receive the event and send back it's own:
```
fullLatency = Date.now() - time1
oneWayLatency = fullLatency / 2.
```
Client can also get the difference in system time compared to the server:
```
timeOffset = Date.now() - time2 - oneWayLatency
```
So now when the server sends a timestamp the client can translate it to the client's system time:
```
fixedTimestamp = timestamp + timeOffset
```

As latency can be variable this process is repeated every 2 seconds and averaged over the last 20 requests

### CODE REVIEW

File: front/app/src/js/components/game/Engine/_GameSocketIO.js [hunk 1]
```diff
@@ -44,32 +44,12 @@ export class _GameSocketIO {
...
-        return;
-      } else {
-        console.error('Connection error:', error['message']);
-        ToastNotifications.addErrorNotification(`Connection error: ${error['message']}`);
-        this.disconnect();
-        this.#reconnectAttempts++;
-        await sleep(2000);
-        await this.init(URI);
-      }
+      await this.#connectErrorHandler(jsonString, URI);
```
Comment (V-Fries) [inline front/app/src/js/components/game/Engine/_GameSocketIO.js:52]: Client requests time sync when connected

File: front/app/src/js/components/game/Engine/_GameSocketIO.js [hunk 2]
```diff
@@ -86,167 +66,226 @@ export class _GameSocketIO {
...
+      await sleep(2000);
+      await this.init(URI);
+    }
+  }
+
+  async #sceneEventHandler(data) {
+    while (!ServerTime.isTimeSynced) {
+      await sleep(5);
+    }
+    const scene = new Scene();
```
Comment (V-Fries) [inline front/app/src/js/components/game/Engine/_GameSocketIO.js:138]: Wait for the first time sync before creating the scene as some time related data will need adjusting

File: front/app/src/js/components/game/Engine/_GameSocketIO.js [hunk 3]
```diff
@@ -86,167 +66,226 @@ export class _GameSocketIO {
...
+    const movement = data['movement'];
+    match.setBallMovement(movement);
+    match.setBallPosition(data['position']);
+
+    const paddle = movement['x'] < 0. ?
+        match.players[0].paddle :
+        match.players[1].paddle;
+    const timeAtUpdate = ServerTime.fixServerTime(data['time_at_update']);
+    const now = Number(Date.now());
+    const timeDelta = (now - timeAtUpdate + ServerTime.latency) / 1000.;
```
Comment (V-Fries) [inline front/app/src/js/components/game/Engine/_GameSocketIO.js:207]: Update ball now takes latency into account

File: front/app/src/js/components/game/Engine/_GameSocketIO.js [hunk 4]
```diff
@@ -86,167 +66,226 @@ export class _GameSocketIO {
       console.log('Server debug message: ', message);
     });
 
-    this.#socketIO.on('sync_time', (serverTime) => {
-      ServerTimeFixer.updateServerLatency(serverTime);
+    this.#socketIO.on('time_sync', (timeData) => {
+      ServerTime.syncServerTime(timeData);
     });
```
Comment (V-Fries) [inline front/app/src/js/components/game/Engine/_GameSocketIO.js:72]: New event to sync the time between server and client. Server sends it after the client emits `request_time_sync`

File: front/app/src/js/components/game/Engine/_GameSocketIO.js [hunk 5]
```diff
@@ -9,7 +9,7 @@ import {PlayerLocation} from '../Scene/PlayerLocation.js';
 import {sleep} from '../sleep.js';
```
Comment (V-Fries) [inline front/app/src/js/components/game/Engine/_GameSocketIO.js:9]: This file is full of refactoring so I highlighted the important changes with comments

**Review by tdameros — APPROVED** [review]

**Review by EthanDelage — APPROVED** [review]

**Pull request merged**